File: Question 0113 - Path Sum II/src/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    private void solve(TreeNode root, int targetSum, List<Integer> path, List<List<Integer>> ans, int curSum)
    {
        if(root == null) return;
        path.add(root.val);
        curSum += root.val;
        if(root.left == null && root.right == null && targetSum == curSum) ans.add(new ArrayList<>(path));


        if(root.left != null)solve(root.left, targetSum, path, ans, curSum);
        if(root.right != null) solve(root.right, targetSum, path, ans, curSum);

        path.remove(path.size()-1);

    }
    public List<List<Integer>> pathSum(TreeNode root, int targetSum)
    {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> path = new ArrayList<>();

        solve(root, targetSum, path, ans, 0);
        return ans;
    }
}

File: Question 0113 - Path Sum II/src/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
private:
    void solve(TreeNode *root, vector<vector<int>> &ans, vector<int> &curr, int targetSum)
    {
        if (!root)
            return;

        curr.push_back(root->val);
        if (!root->left && !root->right && targetSum == root->val)
        {
            ans.push_back(curr);
        }
        else
        {
            if (root->left)
            {
                solve(root->left, ans, curr, targetSum - root->val);
            }

            if (root->right)
            {

                solve(root->right, ans, curr, targetSum - root->val);
            }
        }
        curr.pop_back();
    }

public:
    vector<vector<int>> pathSum(TreeNode *root, int targetSum)
    {
        vector<vector<int>> ans;
        vector<int> curr;
        solve(root, ans, curr, targetSum);
        return ans;
    }
};